### ISSUE TITLE
Discussion: A better way to express Group TopN?

### ISSUE THREAD

**Issue by lmatz** [risingwavelabs/risingwave#7519]
Min_by 
Max_by

https://medium.com/snowflake/snowflake-min-by-and-max-by-aggregate-functions-8c0c7f30058e

https://docs.snowflake.com/en/sql-reference/functions/min_by.html

**Comment by TennyZhuang** [comment]
We've discussed this solution before. The idea is similar, we suggested `SELECT (array_agg((c1, c2, c3) order by c1 limit 10)).* FROM t1 GROUP BY c2`, which is even more powerful than `min_by`.

The solution works well for batch, but the overhead may be too high in the streaming engine.

The real difference is the output format, with the `OVER` or `LATERAL JOIN` solution the result is flattened.

```
+--------------+
| ID | STATUS  |
|--------------+
| 101 | STATE2 |
| 101 | STATE3 |
| 102 | STATE7 |
| 102 | STATE6 |
+--------------+
```
 
With the `min_by` or `array_agg` solution, the result is nested.

```
+-------------+-------------+
|          ID | STATUS      |
|-------------+-------------+
|         101 | [           |
|             | STATE2,     |
|             | STATE3      |
|             | ]           |
|         102 | [           |
|             | STATE7,     |
|             | STATE6      |
|             | ]           |
+-------------+-------------+
```

In our stream engine, we only support row-level operations, which means we have to output the entire array update when inserting a record, e.g:

```
Op::Update(
   (101, [STATE2, STATE3],
   (101, [STATE2, STATE8],
```

In the flattened case, we only need to output two records:

```
Op::Delete(101, STATE3)
Op::Insert(101, STATE8)
```

The nested solution introduces a group size factor into the time complexity.

**Comment by TennyZhuang** [comment]
TL;DR
It's hard for the stream engine to represent the incremental updates of the nested types.
